File: 08-03-fragment-programming/java-08-03-fragment-programming/app/src/main/java/com/example/fragment/FirstFragment.java
package com.example.fragment;

import android.content.Context;
import android.os.Bundle;
import android.support.v4.app.Fragment;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

public class FirstFragment extends Fragment {

    private OnButtonClickedListener mOnButtonClickedListener;

    public FirstFragment() {
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        if (context instanceof OnButtonClickedListener) {
            mOnButtonClickedListener = (OnButtonClickedListener) context;
        }
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_first, container, false);
        Button helloButton = view.findViewById(R.id.hello_button);
        helloButton.setOnClickListener(v -> mOnButtonClickedListener.onButtonClicked());
        return view;
    }

    @Override
    public void onDetach() {
        super.onDetach();
        mOnButtonClickedListener = null;
    }

    public interface OnButtonClickedListener {
        void onButtonClicked();
    }
}

File: 08-03-fragment-programming/kotlin-08-03-fragment-programming/app/src/main/java/com/example/fragment/FirstFragment.kt
package com.example.fragment

import android.content.Context
import android.os.Bundle
import android.support.v4.app.Fragment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button

class FirstFragment : Fragment() {

    private var mOnButtonClickedListener: OnButtonClickedListener? = null

    override fun onAttach(context: Context?) {
        super.onAttach(context)
        if (context is OnButtonClickedListener) {
            mOnButtonClickedListener = context
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_first, container, false)
        val helloButton = view.findViewById<Button>(R.id.hello_button)
        helloButton.setOnClickListener { v -> mOnButtonClickedListener?.onButtonClicked() }
        return view
    }

    override fun onDetach() {
        super.onDetach()
        mOnButtonClickedListener = null
    }

    interface OnButtonClickedListener {
        fun onButtonClicked()
    }
}
